common/dev/xdpe15284.c

```diff
@@ -22,22 +22,139 @@
 #include "util_pmbus.h"
 #include "pmbus.h"
+#include "xdpe15284.h"
 
 LOG_MODULE_REGISTER(xdpe15284);
 
 #define XDPE15284_MFR_DISABLE_SECURITY_ONCE 0xCB
 #define XDPE15284_MFR_FW_CMD_REG 0xFE
 #define XDPE15284_MFR_FW_CMD_DATA_REG 0xFD
-#define XDPE15284_MFR_FW_CMD_DATA_LEN 5
+#define XDPE15284_MFR_FW_CMD_DATA_LEN 4
+#define XDPE15284_REMAINING_WRITE_DATA_LEN 2
 #define XDPE15284_NOP_CMD 0x00
+#define XDPE15284_GET_REMAINING_WR_CMD 0x10
 #define XDPE15284_CALCULATE_CRC_CMD 0x2D
-#define XDPE15284_CALCULATE_CRC_DELAY_MS 10
+#define XDPE15284_WAIT_DATA_DELAY_MS 10
+#define XDPE15284C_CONF_SIZE 1344 // Config(604) + PMBus(504) + SVID(156) + PMBusPartial(80)
+
+#define XDPE15284_WRITE_PROTECT_CMD_REG 0x10
+#define XDPE15284_WRITE_PROTECT_DEFAULT_VAL 0xFF
 
 const uint32_t REG_LOCK_PASSWORD = 0x7F48680C;
+static uint8_t write_protect_default_val = XDPE15284_WRITE_PROTECT_DEFAULT_VAL;
 
-bool xdpe15284_get_checksum(uint8_t bus, uint8_t addr, uint8_t *checksum)
+static bool xdpe15284_set_write_protect_reg(uint8_t bus, uint8_t addr, uint8_t optional)
 {
-	CHECK_NULL_ARG_WITH_RETURN(checksum, false);
+	uint8_t set_val = 0;
+
+	switch (optional) {
+	case XDPE15284_ENABLE_WRITE_PROTECT:
+		set_val = write_protect_default_val;
+		break;
+	case XDPE15284_DISABLE_WRITE_PROTECT:
+		set_val = XDPE15284_DISABLE_WRITE_PROTECT_VAL;
+		break;
+	default:
+		LOG_ERR("Invalid optional: 0x%x to set write protect reg", optional);
+		return false;
+	}
+
+	uint8_t retry = 3;
+	I2C_MSG i2c_msg = { 0 };
+
+	i2c_msg.bus = bus;
+	i2c_msg.target_addr = addr;
+	i2c_msg.tx_len = 2;
+	i2c_msg.data[0] = XDPE15284_WRITE_PROTECT_CMD_REG;
+	i2c_msg.data[1] = set_val;
+
+	if (i2c_master_write(&i2c_msg, retry)) {
+		LOG_ERR("Set write protect register fail, bus: 0x%x, addr: 0x%x, set_val: 0x%x",
+			bus, addr, set_val);
+		return false;
+	}
+
+	k_msleep(XDPE15284_WAIT_DATA_DELAY_MS);
+
+	memset(&i2c_msg, 0, sizeof(I2C_MSG));
+	i2c_msg.bus = bus;
+	i2c_msg.target_addr = addr;
+	i2c_msg.tx_len = 1;
+	i2c_msg.rx_len = 1;
+	i2c_msg.data[0] = XDPE15284_WRITE_PROTECT_CMD_REG;
+
+	if (i2c_master_read(&i2c_msg, retry)) {
+		LOG_ERR("Read write protect register fail, bus: 0x%x, addr: 0x%x", bus, addr);
+		return false;
+	}
+
+	if (i2c_msg.data[0] != set_val) {
+		LOG_ERR("Set write protect register fail, bus: 0x%x, addr: 0x%x, ret_val: 0x%x, set_val: 0x%x",
+			bus, addr, i2c_msg.data[0], set_val);
+		return false;
+	}
+
+	return true;
+}
+
+static bool init_write_protect_default_val(uint8_t bus, uint8_t addr)
+{
+	uint8_t retry = 3;
+	I2C_MSG i2c_msg = { 0 };
+
+	i2c_msg.bus = bus;
+	i2c_msg.target_addr = addr;
+	i2c_msg.tx_len = 1;
+	i2c_msg.rx_len = 1;
+	i2c_msg.data[0] = XDPE15284_WRITE_PROTECT_CMD_REG;
+
+	if (i2c_master_read(&i2c_msg, retry)) {
+		LOG_ERR("Read write protect register fail, bus: 0x%x, addr: 0x%x", bus, addr);
+		return false;
+	}
+
+	write_protect_default_val = i2c_msg.data[0];
+	return true;
+}
+
+bool xdpe15284_enable_write_protect(uint8_t bus, uint8_t addr)
```

Comment:
The code for xdpe15284_enable_write_protect and xdpe15284_disable_write_protect looks exactly same to me. I think its better to have a single API xdpe15284_set_write_protect with enable/disable option so u can pass it to xdpe15284_set_write_protect_reg, isn't it?

Reply:
Hi @jagpalgill 
Modify with your suggestions, please help to review this commit again, thank you.